### PR TITLE
sclang: fix non-local return

### LINKED ISSUE #232: 'try { loop ... }' doesn't unwind the stack properly if an error occurs within the loop
[Issue migrated from SourceForge | ID: 3472743 | Submitted by 'jamshark70']
[http://sourceforge.net/support/tracker.php?aid=3472743]

```
x = [nil];
"%\n".postf(x.do({ |item|
try { max(1, x) == 2 }
}));
```

"try" should swallow the error, and the postf call should print the value of x (the return from 'do'). Instead:

```
ERROR: Message 'postf' not understood.
RECEIVER:
Instance of Array {    (0x48a1e78, gc=E0, fmt=01, flg=00, set=02)
  indexed slots [0]
}
ARGS:
Instance of Array {    (0x4831508, gc=E4, fmt=01, flg=00, set=02)
  indexed slots [1]
      0 : nil
}
```

The empty array that is the receiver of postf is postf's default argument -- see String.findMethod(\postf).prototypeFrame.

The most likely explanation is that some function or method arguments get pushed onto the stack, and they are supposed to be removed before postf, but the combination of 'try' and the loop causes this not to happen.
- It seems to occur only if 'try' is inside a loop.

```
"%\n".postf(try { max(1, nil) == 1 })  // postf succeeds
```
- The bug does NOT occur if the method applied to the loop result has no arguments (receiver only). I discovered the issue when trying to do ["First item"] ++ anArray.select { try { ... } { false } }.

```
// 'size' succeeds
x.select({ |item|
try { max(1, x) == 2 } { false }
}).size
```
- The issue does not occur in a five month old revision. I bet the bug was there ever since try/protect were introduced. Since it wasn't discovered in several years (!) it's not likely to cause major headaches in 3.5, so don't waste time trying to fix this for 3.5.

Marked for 3.6.

### LINKED ISSUE #7448: Lang Crashes periodically with what looks like a GC issue
I've seen a few of these now with crashes at ```AllocPool::Alloc(unsigned long)``` or in ```PyrGC```. Below are three dumps from yesterday. EDIT: Now have a reproducer

## Environment

* SuperCollider version: 3.14.1
* Operating system: MacOS 15.7.4
* Other details (Qt version, audio driver, etc.): Qt 6.7.3

## Steps to reproduce
This crashes pretty reliably. Boiled it down from some more complex code I was running.

```(
var bufCol, count = 0;

bufCol = IdentityDictionary();

SkipJack({
	count.postln;
	count = count + 1;
	100.do({try { bufCol[Pdef(\doesNotExist).source[\buf].source.bufnum] } });
}, 0.001);
)
```

## Expected vs. actual behavior

Crash dumps: 

https://gist.github.com/muellmusik/e77cce4afaf0053e335445e3e589145f
https://gist.github.com/muellmusik/312fefb54c5d01ff37b28c2005c4d01a
https://gist.github.com/muellmusik/ef1d76ca28ecd462d2380cab7c962283

### PR DESCRIPTION
**Pull request by JordanHendersonMusic**
## Purpose and Motivation

Non local returns don't unwind the stack.

Closes #7448

Closes #232  (I think...)

## Types of changes

- Bug fix
- Breaking change? I don't think so, but technically people might have relied on the borked behaviour.

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] This PR is ready for review

### CODE REVIEW

**Comment by dyfer** [comment]
This is a huge improvement, thanks @JordanHendersonMusic for taking this on!

> Closes #232 (I think...)

Yes, it seems so! At the very least, the reproducer from @telephon now functions as expected: 
```supercollider
x = [nil]; "hello".postf(try { max(x, 1) }); // hello-> hello
```

This is way beyond my understanding for a proper review, but it has been such a long-standing issue and I wanted to express my appreciation for the fix.

File: testsuite/classlibrary/TestNonLocalReturn.sc [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		var r = try { nil.(1, 2, 3, this.tail_call_that_throws_with_things_on_stack) };
+
+		this.assertEquals(r, nil);
+
+		after = this.stackDepth;
+		this.assertEquals(before, after, "Non local returns should not leave things on the stack");
+	}
+
+
+	tail_call{ ^1 + 1 }
```
Comment (dyfer) [inline testsuite/classlibrary/TestNonLocalReturn.sc]:
```suggestion
	tail_call { ^1 + 1 }
```

File: testsuite/classlibrary/TestNonLocalReturn.sc [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	tail_call_that_throws { ^Error("throwing").throw }
+	tail_call_that_throws_with_things_on_stack {
+		// Yes this is weird because 1 is on the stack.
+		// As of 2026, Function.forkIfNeeded does this.
```
Comment (dyfer) [inline testsuite/classlibrary/TestNonLocalReturn.sc]:
```suggestion
		// As of SC 3.15, Function.forkIfNeeded does this.
```

**Review by dyfer — COMMENTED** [review]
I just have a couple of minor/formatting suggestions, feel free to disregard

**Review by JordanHendersonMusic — COMMENTED** [review]
The changes to the allocator method is something @Spacechild1 and I did, it's just copy and pasted from another method. It's a bug I found when doing this.

**Comment by JordanHendersonMusic** [comment]
> This is way beyond my understanding for a proper review, but it has been such a long-standing issue and I wanted to express my appreciation for the fix.

Thanks!

To be honest, there isn't much to review. What happened before was wrong. If you see here https://github.com/supercollider/supercollider/issues/7448#issuecomment-4243038009 I've described exactly what is going wrong.

Really it's a question of whether the tests are adequate and capture all the edge cases. I think I need to add some that return from the block rather than the method. Right now I'm only applying the fix when you return from a method - which I think makes sense, but should check with tests! It might also be good to test yield, and perhaps some custom non-local return use. 

This does open up the possibility of using non-local returns more generally, there are some bizarre things you can do!

**Comment by dyfer** [comment]
> To be honest, there isn't much to review. 

...maybe, if one understands how sclang works internally :)

> Really it's a question of whether the tests are adequate and capture all the edge cases.

Just in case it's related: we also have an issue when e.g. manipulating large ~~dictionaries~~ lists. I don't think I ever filed it because I don't exactly understand what's going on... seems to be related to GC: when the dictionary is above certain size and its contents are being manipulated, it seems to be getting other data... See code and comments below:

<details>

```supercollider
(
var allItems;
var numItems = 1000; // numbers below 300 will likely not cause the error

allItems = numItems.collect({ Dictionary() }).as(List);

// ~allItems = allItems; // !!!!! uncommenting this gets rid of the error...

allItems.do({|thisDict, inc|
	var newDict;

	newDict = Dictionary.newFrom((\inc: inc));
	allItems.add(newDict); // add to the List

	allItems.do({|t|
		if(t.isKindOf(Dictionary).not) {
			"%: detected a non-dictionary entry after add: %".format(inc, t.asCompileString).error;
			// this.halt;
		};
	});

	allItems.remove(thisDict); // current item from the list
	allItems.do({|t|
		if(t.isKindOf(Dictionary).not) {
			"%: detected a non-dictionary entry after remove: %".format(inc, t.asCompileString).error;
			// this.halt;
		};
	});
});
)
```

</details>

**Comment by JordanHendersonMusic** [comment]
There's an issue for that one, was just looking at it!

Yeah... I have no idea what is happening there. I doubt it's GC related though. I think it's stack related but can't find where it goes wrong and why the size has something to do with it.

**Comment by lnihlen** [comment]
Have you tested the performance overhead of this? Otherwise LGTM

File: lang/LangSource/PyrInterpreter3.cpp [hunk 3]
```diff
@@ -511,14 +512,24 @@ static inline void checkStackDepth(VMGlobals* g, PyrSlot* sp) {
 #    pragma GCC optimize("-fno-gcse")
 #endif
 
+// Enable this to print each opcode and the stack size
```
Comment (lnihlen) [inline lang/LangSource/PyrInterpreter3.cpp]: Do you want to leave this in? Should it be in a separate PR?

**Review by lnihlen — APPROVED** [review]
Overall LGTM. Let me know if you want help benchmarking this change with LangBench.

File: lang/LangSource/PyrInterpreter3.cpp (hunk 3 above)
Comment (JordanHendersonMusic) [reply lang/LangSource/PyrInterpreter3.cpp]: Hmm yes should probably be separate because I want to remove all the current print debugging and redo it with a cleaner interface. Most of the old commented-out code doesn't work anymore anyway... Perhaps this could replaced with flags in the build system, or even just a header, just so they are all in once place.

**Comment by JordanHendersonMusic** [comment]
Thanks! No I haven't yet tested it. I'll try your bench some time this week and get back to you on that

**Review by telephon — APPROVED** [review]

**Comment by JordanHendersonMusic** [comment]
@lnihlen Well I ran the LangBench... and assuming number_of_runs wants to be high, then I'm confused because either this improves performance by 100%, or something else is happening...

```
// develop
--- start of json ---
{
"awfy.macro.richards": {
        "wall_clock_time": 60.050890193,
        "number_of_runs": 1017
},
"awfy.micro.queens": {
        "wall_clock_time": 60.001482897999,
        "number_of_runs": 13323
},
"awfy.micro.bounce": {
        "wall_clock_time": 60.000706273,
        "number_of_runs": 38802
},
"awfy.micro.permute": {
        "wall_clock_time": 60.004194268999,
        "number_of_runs": 7405
},
"awfy.macro.havlak": {
        "wall_clock_time": 61.722593091,
        "number_of_runs": 16
},
"awfy.micro.storage": {
        "wall_clock_time": 60.002134415003,
        "number_of_runs": 7009
},
"awfy.micro.mandelbrot": {
        "wall_clock_time": 60.000003883042,
        "number_of_runs": 16737422
},
"awfy.micro.nbody": {
        "wall_clock_time": 60.000013532021,
        "number_of_runs": 2726825
},
"awfy.macro.cd": {
        "wall_clock_time": 60.45111086,
        "number_of_runs": 150
},
"awfy.micro.sieve": {
        "wall_clock_time": 60.003033341003,
        "number_of_runs": 15213
},
"awfy.micro.list": {
        "wall_clock_time": 60.000893714004,
        "number_of_runs": 52184
},
"awfy.micro.towers": {
        "wall_clock_time": 60.004773875,
        "number_of_runs": 13270
},
"awfy.macro.deltablue": {
        "wall_clock_time": 60.006101054997,
        "number_of_runs": 6352
}
}
--- end of json ---
```

```
// this pr
--- start of json ---
{
"awfy.macro.richards": {
        "wall_clock_time": 60.014470259,
        "number_of_runs": 1003
},
"awfy.micro.queens": {
        "wall_clock_time": 60.001484546003,
        "number_of_runs": 24307
},
"awfy.micro.bounce": {
        "wall_clock_time": 60.000311722996,
        "number_of_runs": 64244
},
"awfy.micro.permute": {
        "wall_clock_time": 60.000221626998,
        "number_of_runs": 14643
},
"awfy.macro.havlak": {
        "wall_clock_time": 61.082240919,
        "number_of_runs": 26
},
"awfy.micro.storage": {
        "wall_clock_time": 60.000504857,
        "number_of_runs": 12328
},
"awfy.micro.mandelbrot": {
        "wall_clock_time": 60.000001458032,
        "number_of_runs": 26579729
},
"awfy.micro.nbody": {
        "wall_clock_time": 60.000006117006,
        "number_of_runs": 4262146
},
"awfy.macro.cd": {
        "wall_clock_time": 60.115723143,
        "number_of_runs": 336
},
"awfy.micro.sieve": {
        "wall_clock_time": 60.001248423004,
        "number_of_runs": 36473
},
"awfy.micro.list": {
        "wall_clock_time": 60.000000225001,
        "number_of_runs": 62825
},
"awfy.micro.towers": {
        "wall_clock_time": 60.003460929996,
        "number_of_runs": 12470
},
"awfy.macro.deltablue": {
        "wall_clock_time": 60.004091739,
        "number_of_runs": 12231
}
}
--- end of json ---

```

**Comment by muellmusik** [comment]
Thanks @JordanHendersonMusic so much for fixing this longstanding bug! 🙏
